kong/conf_loader/init.lua

```diff
@@ -396,65 +396,122 @@ local function load(path, custom_conf, opts)
 
     loaded_vaults = setmetatable(vaults, conf_constants._NOP_TOSTRING_MT)
 
-    if get_phase() == "init" then
-      local secrets = getenv("KONG_PROCESS_SECRETS")
-      if secrets then
-        unsetenv("KONG_PROCESS_SECRETS")
-
-      else
-        local path = pl_path.join(abspath(ngx.config.prefix()), unpack(conf_constants.PREFIX_PATHS.kong_process_secrets))
-        if exists(path) then
-          secrets, err = pl_file.read(path, true)
-          pl_file.delete(path)
-          if not secrets then
-            return nil, fmt("failed to read process secrets file: %s", err)
+    -- collect references
+    local is_reference = require "kong.pdk.vault".is_reference
+    for k, v in pairs(conf) do
+      local typ = (conf_constants.CONF_PARSERS[k] or {}).typ
+      v = parse_value(v, typ == "array" and "array" or "string")
+      if typ == "array" then
+        local found
+
+        for i, r in ipairs(v) do
+          if is_reference(r) then
+            found = true
+            if not refs then
+              refs = setmetatable({}, conf_constants._NOP_TOSTRING_MT)
+            end
+            if not refs[k] then
+              refs[k] = {}
+            end
+            refs[k][i] = r
           end
         end
+
+        if found then
+          conf[k] = v
+        end
+
+      elseif is_reference(v) then
+        if not refs then
+          refs = setmetatable({}, conf_constants._NOP_TOSTRING_MT)
+        end
+        refs[k] = v
       end
+    end
 
-      if secrets then
-        secrets, err = process_secrets.deserialize(secrets, path)
+    local prefix = abspath(conf.prefix or ngx.config.prefix())
+    local secret_env
+    local secret_file
+    local secrets = getenv("KONG_PROCESS_SECRETS")
+    if secrets then
+      secret_env = "KONG_PROCESS_SECRETS"
+
+    else
+      local secrets_path = pl_path.join(prefix, unpack(conf_constants.PREFIX_PATHS.kong_process_secrets))
+      if exists(secrets_path) then
+        secrets, err = pl_file.read(secrets_path, true)
         if not secrets then
-          return nil, err
+          pl_file.delete(secrets_path)
+          return nil, fmt("failed to read process secrets file: %s", err)
         end
+        secret_file = secrets_path
+      end
+    end
 
-        for k, deref in pairs(secrets) do
-          local v = parse_value(conf[k], "string")
-          if refs then
-            refs[k] = v
-          else
-            refs = setmetatable({ [k] = v }, conf_constants._NOP_TOSTRING_MT)
-          end
+    if secrets then
+      local env_path = pl_path.join(prefix, unpack(conf_constants.PREFIX_PATHS.kong_env))
+      secrets, err = process_secrets.deserialize(secrets, env_path)
+      if not secrets then
+        return nil, err
+      end
+
+      -- TODO: remember!
```

Comment:
@bungle Do you still remember what is the exact todo item here? I'm working to finish this PR